### ISSUE TITLE
Recognize Pydantic validator decorators as implicit classmethods

### ISSUE THREAD

**Issue by rayzeller** [astral-sh/ty#2719]
## Summary

Pydantic's `@field_validator`, `@model_validator`, and `@field_serializer` decorators implicitly wrap the decorated function as a classmethod. The first parameter receives the class (`cls`), not an instance (`self`).

Currently ty treats these methods as regular instance methods, causing incorrect type inference for the first parameter.

## Example

```python
from pydantic import BaseModel, field_validator

class User(BaseModel):
    name: str

    @field_validator("name")
    def validate_name(cls, v: str) -> str:
        return v.strip()
```

**Current behavior:** `cls` is typed as `User` (instance)  
**Expected behavior:** `cls` is typed as `type[User]` (class)

### Pydantic Documentation
- [Pydantic Validators](https://docs.pydantic.dev/latest/concepts/validators/) - documents that validators receive `cls`

### Prior Art
- **mypy plugin**: [`pydantic/mypy.py`](https://github.com/pydantic/pydantic/blob/main/pydantic/mypy.py) defines `IMPLICIT_CLASSMETHOD_DECORATOR_FULLNAMES` to handle this
- **mypy plugin PR**: [pydantic/pydantic#4102](https://github.com/pydantic/pydantic/pull/4102)
- **pyright**: [microsoft/pyright#804](https://github.com/microsoft/pyright/issues/804) - no special support; recommends explicit `@classmethod`

## Parent Issue

Part of #2403

**Comment by sharkdp** [comment]
Thank you. I added it as a sub-issue of #2403 for now, but before someone goes ahead and implements this, I think there needs to be a larger design discussion in #2403 (for example: how do we even recognize `pydantic` as a dependency to special-case our behavior? do we simply go by name of the module? do we want a config setting to opt into this or should it be enabled by default? which pydantic versions do we want to support? …)

**Comment by sharkdp** [comment]
So while reviewing the implementation in https://github.com/astral-sh/ruff/pull/23272, I noticed that Pydantic explicitly states in [its documentation](https://docs.pydantic.dev/2.7/concepts/validators/) (emphasis mine):

> `@field_validator`s are "class methods", so the first argument value they receive is the `UserModel` class, not an instance of `UserModel`. **We recommend you use the `@classmethod` decorator on them below the `@field_validator` decorator to get proper type checking.**

and:

> Before model validators should be class methods. The first argument should be `cls` (**and we also recommend you use `@classmethod` below `@model_validator` for proper type checking**), …

They also consistently use `@classmethod` in their examples:

> 
> ```py
> class Model(BaseModel):
>     x: int
> 
>     @field_validator('x')
>     @classmethod
>     def validate_x(cls, v: int) -> int:
>         ...
> ```

Given this, I'd like to close this issue — and the implementation in https://github.com/astral-sh/ruff/pull/23272. Instead, I'd like to ask pydantic users to follow established practices.

Let me know if you disagree and we can reconsider opening this.

**Comment by sharkdp** [comment]
I also added some tests in https://github.com/astral-sh/ruff/pull/23556 to make sure that we understand the recommended pattern.
